File: app/src/main/java/com/aavens/mindloft/ui/webpage/WebPageActivity.java
package com.aavens.mindloft.ui.webpage;

import android.content.Intent;
import android.net.Uri;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.aavens.mindloft.R;

import java.net.URL;

public class WebPageActivity extends AppCompatActivity {
    public static final String URL_EXTRA = "com.aavens.mindloft.ui.webpage.URL_EXTRA";

    private WebView webView;
    private String url;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_web_page);
        initToolbar();
        initWebView();
        loadUrl();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.web_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case android.R.id.home:
                onBackPressed();
                break;
            case R.id.reload:
                webView.reload();
                break;
            case R.id.in_browser:
                openLinkInBrowser();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void initToolbar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.default_toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setTitle(R.string.link_title);
    }

    private void initWebView() {
        webView = (WebView) findViewById(R.id.web_view);
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                updateTitle();
            }
        });
    }

    private void updateTitle() {
        getSupportActionBar().setTitle(webView.getTitle());
    }

    private void loadUrl() {
        url = getIntent().getStringExtra(URL_EXTRA);
        if (!url.startsWith("http://") || !url.startsWith("https://")) {
            url = "http://" + url;
        }
        webView.loadUrl(url);
    }

    private void openLinkInBrowser() {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(url));
        startActivity(intent);
    }
}

File: app/src/main/java/com/aavens/mindloft/ui/webpage/WebPageActivity.kt
package com.aavens.mindloft.ui.webpage

import android.content.Intent
import android.net.Uri
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.webkit.WebView
import android.webkit.WebViewClient

import com.aavens.mindloft.R

class WebPageActivity: AppCompatActivity() {

    private var webView: WebView? = null
    private var url: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_web_page)
        initToolbar()
        initWebView()
        loadUrl()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.web_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            android.R.id.home -> onBackPressed()
            R.id.reload -> webView?.reload()
            R.id.in_browser -> openLinkInBrowser()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun initToolbar() {
        val toolbar = findViewById(R.id.default_toolbar) as Toolbar?
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setTitle(R.string.link_title)
    }

    private fun initWebView() {
        webView = findViewById(R.id.web_view) as WebView?
        val settings = webView?.settings
        settings?.javaScriptEnabled = true
        webView?.setWebViewClient(object : WebViewClient() {
            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                updateTitle()
            }
        })
    }

    private fun updateTitle() {
        supportActionBar?.title = webView?.title
    }

    private fun loadUrl() {
        url = intent.getStringExtra(URL_EXTRA)
        webView?.loadUrl(url)
    }

    private fun openLinkInBrowser() {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(url)
        startActivity(intent)
    }

    companion object {
        val URL_EXTRA = "com.aavens.mindloft.ui.webpage.URL_EXTRA"
    }
}
